### ISSUE TITLE
Server run() method performs SystemExit(1) instead of throwing the actual OSError with throwback

### ISSUE THREAD

**Issue by mahee96** [pallets/werkzeug#2687]
Running Flask development web server on a local Minikube cluster in a pod, the server is configured to fallback to different(known) port when the requested port is IN USE by the other replica container.

The implementation for this involves catching the SocketException or OSError in python, but on checking the code path, seems the code in werkzeug/serving.py->BaseWSGIServer->__init__() method **intentionally** consumes the OSError and throwback and raises SystemExit(1). 

This doesn't help since, the code if catches the SystemExit, then it has to inspect what was the exit for, and other reasons, instead of catching the OSError and matching with ErrorNo for SocketException.

This is required for arbitration as per my current implementation.

In this repo:
https://github.com/pallets/werkzeug/blob/825633eb783f49b0450b0e6866bc55985b154531/src/werkzeug/serving.py#L750-L774

custom server implementation snippet:
```python
...

import socket, errno
import sys, traceback

if __name__ == '__main__':
    try:
        server.run(host="0.0.0.0", port=5000, debug=False)
    # except socket.error as serr:
    #     print(err)
    # except OSError as serr:
    #     print(err, err == type)
    except SystemExit as e:
        print(e.__class__.__name__)
        traceback.print_exc(file=sys.stdout)
        server.run(host="0.0.0.0", port=6000, debug=False)
    #    raise e

```

![image](https://user-images.githubusercontent.com/47920326/235518028-940a9ab4-eea0-445a-8449-1f18ae890835.png)

Expectation:
    throw OSError as is without consuming it

Environment:

- Python version: Python 3.10.11
- Werkzeug version: Werkzeug-2.2.3.

**Comment by mahee96** [comment]
Thanks for the response, I have a workaround for this though, but was interested in knowing why the actual exception was completely engulfed and system.exit was called.

```python
from flask import Flask
import errno
server = Flask(__name__)

if __name__ == '__main__':
    try:
        server.run(host="0.0.0.0", port=5000, debug=False)
    except SystemExit as e:
        ctx = e.__context__
        if type(ctx) is OSError and ctx.errno == errno.EADDRINUSE:
            server.run(host="0.0.0.0", port=6000, debug=False)
```

I am not planning to use it in production, so no worries :-)

**Comment by davidism** [comment]
Again, what you're doing does not make sense. You already have control of your environment in a container, you do not need to fall back to another port. There are other methods available for finding a free port if you do need to.

**Comment by mahee96** [comment]
I am following a tutorial on microservices (video to Audio converter service) where the tutor creates a pod for each service like auth, gateway, converter, notification.

If I increase the replicas of deployment of auth service it scales in pod level, but I `wanted to try multiple container in same pod`, so I tried with this implementation where the same auth service will now use _same docker image_ having _same codebase_ for both the containers of auth service, which belong to same pod (auth-service pod).

For this I `needed the code to be unmodified between the containers` (no docker rebuild/config changes except exposing new port and adding new container in the deployment.yaml), so I had to `try for a quick hack where the server code will fallback to another known port when requested port has been taken over by its sibling container`. 

I couldn't catch the OSError/socket.error upon startup of the server when it complains of "Address already in use" using simple `except OSError` `except Exception` handlers.

Found out that SystemExit is derived directly from BaseException and hence had to get the __context__ to get the previous exception (socket.error) in the SILENT system exit exception chain.

I understand that this isn't a proper use case, but as for my learning wanted to try this quick hack out, when I faced this missing stack trace issue and wanted to know why it was engulfed.
